esmvalcore/preprocessor/_shared.py

```diff
@@ -266,3 +268,305 @@ def wrapper(data: DataType, *args: Any, **kwargs: Any) -> DataType:
+
+    elif _is_lazy_masked_data(array):
+        # broadcast_to strips masks so we need to handle them explicitly.
+        mask = da.ma.getmaskarray(array)
+        new_mask = broadcast(mask)
+        new_array = da.ma.masked_array(new_array, new_mask)
+
+    return new_array
+
+
+def _is_lazy_masked_data(array):
+    """Similar to `iris._lazy_data.is_lazy_masked_data`."""
+    return isinstance(array, da.Array) and isinstance(
+        da.utils.meta_from_array(array), np.ma.MaskedArray)
+
+
+def get_weights(
+    cube: Cube,
+    coords: Iterable[Coord] | Iterable[str],
+) -> np.ndarray | da.Array:
+    """Calculate suitable weights for given coordinates."""
+    npx = get_array_module(cube.core_data())
+    weights = npx.ones_like(cube.core_data())
+
+    # Time weights: lengths of time interval
+    if 'time' in coords:
+        weights *= broadcast_to_shape(
+            npx.array(get_time_weights(cube)),
+            cube.shape,
+            cube.coord_dims('time'),
+        )
+
+    # Latitude weights: cell areas
+    if 'latitude' in coords:
+        cube = cube.copy()  # avoid overwriting input cube
+        if (
+                not cube.cell_measures('cell_area') and
+                not cube.coords('longitude')
+        ):
+            raise CoordinateNotFoundError(
+                f"Cube {cube.summary(shorten=True)} needs a `longitude` "
+                f"coordinate to calculate cell area weights for weighted "
+                f"distance metric over coordinates {coords} (alternatively, "
+                f"a `cell_area` can be given to the cube as supplementary "
+                f"variable)"
+            )
+        try_adding_calculated_cell_area(cube)
+        weights *= broadcast_to_shape(
+            cube.cell_measure('cell_area').core_data(),
+            cube.shape,
+            cube.cell_measure_dims('cell_area'),
+        )
+
+    return weights
+
+
+def get_time_weights(cube: Cube) -> np.ndarray | da.core.Array:
+    """Compute the weighting of the time axis.
+
+    Parameters
+    ----------
+    cube:
+        Input cube.
+
+    Returns
+    -------
+    np.ndarray or da.Array
+        Array of time weights for averaging. Returns a
+        :class:`dask.array.Array` if the input cube has lazy data; a
+        :class:`numpy.ndarray` otherwise.
+
+    """
+    time = cube.coord('time')
+    coord_dims = cube.coord_dims('time')
+
+    # Multidimensional time coordinates are not supported: In this case,
+    # weights cannot be simply calculated as difference between the bounds
+    if len(coord_dims) > 1:
+        raise ValueError(
+            f"Weighted statistical operations are not supported for "
+            f"{len(coord_dims):d}D time coordinates, expected 0D or 1D"
+        )
+
+    # Extract 1D time weights (= lengths of time intervals)
+    time_weights = time.lazy_bounds()[:, 1] - time.lazy_bounds()[:, 0]
+    if not cube.has_lazy_data():
+        time_weights = time_weights.compute()
+    return time_weights
```

Comment:
I think you missed the changes of https://github.com/ESMValGroup/ESMValCore/pull/2404 here. I will re-do the merge to make sure nothing else is missing

Reply:
Hmm I merged the latest main as it was yesterday, no cherrypicking mor rebase - could git have missed it via a simple merge?

Reply:
Bouwe changed the lines in the `_time.py`, and I moved the original function (w/o Bouwe's changes) to `_shared.py`. Thus, simply removing the function from the time module is not enough, you also need to move that to shared. Will do that now.

Reply:
Ouch! Indeed that then - didn't occur to me I should check the consistency of the function, sorry, bud. At least I was well verbose about what I did 😁